Задание 1. Загрузка данных
Изучить представленный набор данных на основе описания его столбцов, загрузить его и оставить 8 столбцов для дальнейшего изучения: surgery?, Age, rectal temperature, pulse, respiratory rate, temperature of extremities, pain, outcome.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('Horses.csv', na_values = '?', header = None)

In [3]:
df.head()

0   1        2     3      4     5    6    7    8    9   ...    18    19  \
0  2.0   1   530101  38.5   66.0  28.0  3.0  3.0  NaN  2.0  ...  45.0   8.4   
1  1.0   1   534817  39.2   88.0  20.0  NaN  NaN  4.0  1.0  ...  50.0  85.0   
2  2.0   1   530334  38.3   40.0  24.0  1.0  1.0  3.0  1.0  ...  33.0   6.7   
3  1.0   9  5290409  39.1  164.0  84.0  4.0  1.0  6.0  2.0  ...  48.0   7.2   
4  2.0   1   530255  37.3  104.0  35.0  NaN  NaN  6.0  2.0  ...  74.0   7.4   

    20   21   22  23     24  25  26  27  
0  NaN  NaN  2.0   2  11300   0   0   2  
1  2.0  2.0  3.0   2   2208   0   0   2  
2  NaN  NaN  1.0   2      0   0   0   1  
3  3.0  5.3  2.0   1   2208   0   0   1  
4  NaN  NaN  2.0   2   4300   0   0   2  

[5 rows x 28 columns]

In [4]:
df.columns = ['surgery?', 'Age', 'Hospital Number', 'rectal temperature', 'pulse', 'respiratory rate', 'temperature of extremities', 'peripheral pulse', 'mucous membranes', 'capillary refill time', 'pain', 'peristalsis', 'abdominal distension', 'nasogastric tube', 'nasogastric reflux', 'nasogastric reflux PH', 'rectal examination - feces', 'abdomen', 'packed cell volume', 'total protein', 'abdominocentesis appearance', 'abdomcentesis total protein', 'outcome', 'surgical lesion?', 'type of lesion', 'type of lesion', 'type of lesion', 'cp_data']

In [5]:
df.head()

surgery?  Age  Hospital Number  rectal temperature  pulse  \
0       2.0    1           530101                38.5   66.0   
1       1.0    1           534817                39.2   88.0   
2       2.0    1           530334                38.3   40.0   
3       1.0    9          5290409                39.1  164.0   
4       2.0    1           530255                37.3  104.0   

   respiratory rate  temperature of extremities  peripheral pulse  \
0              28.0                         3.0               3.0   
1              20.0                         NaN               NaN   
2              24.0                         1.0               1.0   
3              84.0                         4.0               1.0   
4              35.0                         NaN               NaN   

   mucous membranes  capillary refill time  ...  packed cell volume  \
0               NaN                    2.0  ...                45.0   
1               4.0                    1.0  ...                50.0   
2               3.0                    1.0  ...                33.0   
3               6.0                    2.0  ...                48.0   
4               6.0                    2.0  ...                74.0   

   total protein  abdominocentesis appearance  abdomcentesis total protein  \
0            8.4                          NaN                          NaN   
1           85.0                          2.0                          2.0   
2            6.7                          NaN                          NaN   
3            7.2                          3.0                          5.3   
4            7.4                          NaN                          NaN   

   outcome  surgical lesion?  type of lesion  type of lesion  type of lesion  \
0      2.0                 2           11300               0               0   
1      3.0                 2            2208               0               0   
2      1.0                 2               0               0               0   
3      2.0                 1            2208               0               0   
4      2.0                 2            4300               0               0   

   cp_data  
0        2  
1        2  
2        1  
3        1  
4        2  

[5 rows x 28 columns]

In [6]:
df.drop(['Hospital Number', 'peripheral pulse', 'mucous membranes', 'capillary refill time', 'peristalsis', 'abdominal distension', 'nasogastric tube', 'nasogastric reflux', 'nasogastric reflux PH', 'rectal examination - feces', 'abdomen', 'packed cell volume', 'total protein', 'abdominocentesis appearance', 'abdomcentesis total protein', 'surgical lesion?', 'type of lesion', 'type of lesion', 'type of lesion', 'cp_data'], axis = 1, inplace = True)
df.head()

surgery?  Age  rectal temperature  pulse  respiratory rate  \
0       2.0    1                38.5   66.0              28.0   
1       1.0    1                39.2   88.0              20.0   
2       2.0    1                38.3   40.0              24.0   
3       1.0    9                39.1  164.0              84.0   
4       2.0    1                37.3  104.0              35.0   

   temperature of extremities  pain  outcome  
0                         3.0   5.0      2.0  
1                         NaN   3.0      3.0  
2                         1.0   3.0      1.0  
3                         4.0   2.0      2.0  
4                         NaN   NaN      2.0

Задание 2. Первичное изучение данных
Проанализировать значения по столбцам, рассчитать базовые статистики, найти выбросы.

In [7]:
df.describe()

surgery?         Age  rectal temperature       pulse  \
count  299.000000  300.000000          240.000000  276.000000   
mean     1.397993    1.640000           38.167917   71.913043   
std      0.490305    2.173972            0.732289   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    9.000000           40.800000  184.000000   

       respiratory rate  temperature of extremities       pain     outcome  
count        242.000000                  244.000000  245.00000  299.000000  
mean          30.417355                    2.348361    2.95102    1.551839  
std           17.642231                    1.045054    1.30794    0.737187  
min            8.000000                    1.000000    1.00000    1.000000  
25%           18.500000                    1.000000    2.00000    1.000000  
50%           24.500000                    3.000000    3.00000    1.000000  
75%           36.000000                    3.000000    4.00000    2.000000  
max           96.000000                    4.000000    5.00000    3.000000

Поиск выбросов по столбцу Age:

In [8]:
q1 = df['Age'].quantile(0.25)
q3 = df['Age'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = df[df['Age'].between(lower_bound, upper_bound, inclusive=True)].sort_values('Age')
pd.concat([df, remove_outliers]).drop_duplicates(keep=False)

surgery?  Age  rectal temperature  pulse  respiratory rate  \
3         1.0    9                39.1  164.0              84.0   
9         2.0    9                38.3   90.0               NaN   
13        2.0    9                38.0   92.0              28.0   
16        1.0    9                 NaN  128.0              36.0   
23        1.0    9                38.3  130.0              60.0   
39        1.0    9                39.2  146.0              96.0   
41        2.0    9                39.0  150.0              72.0   
55        1.0    9                38.6  160.0              20.0   
74        1.0    9                 NaN    NaN               NaN   
75        1.0    9                39.7  100.0               NaN   
82        1.0    9                38.1  100.0              80.0   
103       1.0    9                38.0  140.0              68.0   
109       1.0    9                38.4   84.0              40.0   
135       1.0    9                38.1  136.0              48.0   
191       2.0    9                 NaN  100.0              44.0   
212       1.0    9                38.9  120.0              30.0   
227       1.0    9                 NaN    NaN               NaN   
229       1.0    9                38.5  120.0              70.0   
244       1.0    9                38.2  124.0              88.0   
252       1.0    9                38.8  124.0              36.0   
255       1.0    9                38.8  184.0              84.0   
262       2.0    9                38.6   88.0              28.0   
275       1.0    9                38.8  150.0              50.0   
287       2.0    9                39.5   84.0              30.0   

     temperature of extremities  pain  outcome  
3                           4.0   2.0      2.0  
9                           1.0   5.0      1.0  
13                          1.0   1.0      2.0  
16                          3.0   4.0      2.0  
23                          NaN   2.0      1.0  
39                          NaN   NaN      2.0  
41                          NaN   NaN      1.0  
55                          3.0   3.0      2.0  
74                          NaN   NaN      2.0  
75                          3.0   2.0      3.0  
82                          3.0   3.0      1.0  
103                         1.0   3.0      1.0  
109                         3.0   3.0      2.0  
135                         3.0   5.0      2.0  
191                         2.0   4.0      1.0  
212                         1.0   3.0      1.0  
227                         NaN   NaN      1.0  
229                         NaN   NaN      1.0  
244                         1.0   2.0      1.0  
252                         3.0   2.0      2.0  
255                         1.0   4.0      2.0  
262                         NaN   NaN      1.0  
275                         1.0   5.0      2.0  
287                         NaN   NaN      1.0

По столбцу Age выбросом является значение 9, т.к. по описанию в данный столбец должны заносится только значения 1 или 2

Поиск выбросов по столбцу rectal temperature:

In [9]:
q1 = df['rectal temperature'].quantile(0.25)
q3 = df['rectal temperature'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = df[df['rectal temperature'].between(lower_bound, upper_bound, inclusive=True)].sort_values('rectal temperature')
pd.concat([df, remove_outliers]).drop_duplicates(keep=False)

surgery?  Age  rectal temperature  pulse  respiratory rate  \
5         2.0    1                 NaN    NaN               NaN   
7         1.0    1                 NaN   60.0               NaN   
8         2.0    1                 NaN   80.0              36.0   
16        1.0    9                 NaN  128.0              36.0   
20        1.0    1                39.9   72.0              60.0   
..        ...  ...                 ...    ...               ...   
274       1.0    1                 NaN   76.0               NaN   
281       2.0    1                40.0   78.0               NaN   
293       1.0    1                 NaN   78.0              24.0   
295       1.0    1                 NaN  120.0              70.0   
298       1.0    1                36.5  100.0              24.0   

     temperature of extremities  pain  outcome  
5                           2.0   2.0      1.0  
7                           3.0   NaN      2.0  
8                           3.0   4.0      3.0  
16                          3.0   4.0      2.0  
20                          1.0   5.0      1.0  
..                          ...   ...      ...  
274                         NaN   NaN      3.0  
281                         3.0   2.0      2.0  
293                         3.0   NaN      3.0  
295                         4.0   2.0      3.0  
298                         3.0   3.0      1.0  

[61 rows x 8 columns]

Выбросы - значения, ниже, чем значение квантиль 25% = 37.8 и выше, чем квантиль 75% = 38.5, также выведены пропуски

Поиск выбросов по столбцу pulse:

In [10]:
q1 = df['pulse'].quantile(0.25)
q3 = df['pulse'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = df[df['pulse'].between(lower_bound, upper_bound, inclusive=True)].sort_values('pulse')
pd.concat([df, remove_outliers]).drop_duplicates(keep=False)

surgery?  Age  rectal temperature  pulse  respiratory rate  \
3         1.0    9                39.1  164.0              84.0   
5         2.0    1                 NaN    NaN               NaN   
41        2.0    9                39.0  150.0              72.0   
52        2.0    1                 NaN    NaN               NaN   
55        1.0    9                38.6  160.0              20.0   
58        1.0    1                 NaN    NaN              20.0   
74        1.0    9                 NaN    NaN               NaN   
93        2.0    1                 NaN    NaN               NaN   
115       2.0    1                 NaN    NaN              40.0   
117       2.0    1                39.5    NaN               NaN   
126       1.0    1                38.7    NaN               NaN   
151       1.0    1                39.3    NaN               NaN   
159       2.0    1                 NaN    NaN              12.0   
160       1.0    1                 NaN    NaN               NaN   
174       2.0    1                 NaN    NaN               NaN   
204       1.0    1                39.2    NaN              23.0   
216       2.0    1                 NaN    NaN               NaN   
227       1.0    9                 NaN    NaN               NaN   
255       1.0    9                38.8  184.0              84.0   
275       1.0    9                38.8  150.0              50.0   

     temperature of extremities  pain  outcome  
3                           4.0   2.0      2.0  
5                           2.0   2.0      1.0  
41                          NaN   NaN      1.0  
52                          1.0   1.0      1.0  
55                          3.0   3.0      2.0  
58                          4.0   5.0      2.0  
74                          NaN   NaN      2.0  
93                          3.0   5.0      2.0  
115                         3.0   3.0      1.0  
117                         3.0   3.0      3.0  
126                         3.0   5.0      1.0  
151                         4.0   4.0      2.0  
159                         1.0   2.0      1.0  
160                         NaN   4.0      1.0  
174                         3.0   4.0      2.0  
204                         3.0   4.0      1.0  
216                         NaN   NaN      1.0  
227                         NaN   NaN      1.0  
255                         1.0   4.0      2.0  
275                         1.0   5.0      2.0

Выбросы - значения выше, чем квантиль 75% = 88, также выведены пропуски

Поиск выбросов по столбцу respiratory rate:

In [11]:
q1 = df['respiratory rate'].quantile(0.25)
q3 = df['respiratory rate'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = df[df['respiratory rate'].between(lower_bound, upper_bound, inclusive=True)].sort_values('respiratory rate')
pd.concat([df, remove_outliers]).drop_duplicates(keep=False)

surgery?  Age  rectal temperature  pulse  respiratory rate  \
3         1.0    9                39.1  164.0              84.0   
5         2.0    1                 NaN    NaN               NaN   
7         1.0    1                 NaN   60.0               NaN   
9         2.0    9                38.3   90.0               NaN   
29        2.0    1                37.7   48.0               NaN   
..        ...  ...                 ...    ...               ...   
271       1.0    1                37.7   80.0               NaN   
274       1.0    1                 NaN   76.0               NaN   
281       2.0    1                40.0   78.0               NaN   
284       2.0    1                38.5   54.0               NaN   
295       1.0    1                 NaN  120.0              70.0   

     temperature of extremities  pain  outcome  
3                           4.0   2.0      2.0  
5                           2.0   2.0      1.0  
7                           3.0   NaN      2.0  
9                           1.0   5.0      1.0  
29                          2.0   1.0      1.0  
..                          ...   ...      ...  
271                         3.0   5.0      1.0  
274                         NaN   NaN      3.0  
281                         3.0   2.0      2.0  
284                         1.0   3.0      1.0  
295                         4.0   2.0      3.0  

[62 rows x 8 columns]

Выбросы - значения, ниже, чем значение квантиль 25% = 18.5 и выше, чем квантиль 75% = 36, также выведены пропуски

По столбцам surgery?, temperature of extremities, pain - a subjective judgement of the horses pain level, outcome - выбросы отсутствуют, т.к. минимальное и максимальное значения находятся в пределах допустимых значений, будут выведены только отсутствующие данные.

Задание 3. Работа с пропусками
Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе заполнения пропусков по каждому столбцу на основе рассчитанных статистик и возможной взаимосвязи значений в них. Сформировать датафрейм, в котором пропуски будут отсутствовать.

In [12]:
(df.isna().mean() * 100).round()

surgery?                       0.0
Age                            0.0
rectal temperature            20.0
pulse                          8.0
respiratory rate              19.0
temperature of extremities    19.0
pain                          18.0
outcome                        0.0
dtype: float64

Пропуски столбца pain можно восстановить на основании статистики и группировки столбца surgery?, т.к. между уровнем боли лошади и хирургическим вмешательстве возможна взаимосвязь.

In [34]:
df['pain'].fillna(df.groupby('surgery?')['pain'].transform('median'), inplace = True)

Пропуски столбца pulse можно восстановить по значению медианы, т.к. данная метрика = 64.0 отражает отклонение от нормы (30-40 ударов) и значение по медиане не так подвержено выбросам, как метрика mean.

In [35]:
df['pulse'].fillna(df['pulse'].median(), inplace = True)

Пропуски столбца respiratory rate можно восстановить по значению медианы, т.к. данная метрика показывает отклонение от нормальной скорости 8-10 и значение по медиане не так подвержено выбросам, как метрика mean.

In [36]:
df['respiratory rate'].fillna(df['respiratory rate'].median(), inplace = True)

Для столбца rectal temperature можно попробовать сгруппировать по столбцу outcome, для значений outcome 2,3 rectal temperature близка с min либо max значениям, что говорит о состоянии шока (осложнение заболевания).

In [37]:
df['rectal temperature'].fillna(df.groupby('outcome')['rectal temperature'].transform('median'), inplace = True)

Столбец temperature of extremities можно восстановить через группировку по столбцу rectal temperature по медиане, т.к. между показателями существует прямая связь (чем выше rectal temperature, тем выше temperature of extremities). 

In [38]:
df['temperature of extremities'].fillna(df.groupby('rectal temperature')['temperature of extremities'].transform('median'), inplace = True)

In [39]:
(df.isna().mean() * 100).round()

surgery?                      0.0
Age                           0.0
rectal temperature            0.0
pulse                         0.0
respiratory rate              0.0
temperature of extremities    1.0
pain                          0.0
outcome                       0.0
dtype: float64

In [40]:
df.describe()

surgery?         Age  rectal temperature       pulse  \
count  299.000000  300.000000          300.000000  300.000000   
mean     1.397993    1.640000           38.228333   71.280000   
std      0.490305    2.173972            0.692384   27.541545   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.875000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.600000   88.000000   
max      2.000000    9.000000           40.800000  184.000000   

       respiratory rate  temperature of extremities        pain     outcome  
count        300.000000                  298.000000  300.000000  299.000000  
mean          29.273333                    2.364094    2.900000    1.551839  
std           16.010979                    1.014617    1.203395    0.737187  
min            8.000000                    1.000000    1.000000    1.000000  
25%           20.000000                    1.000000    2.000000    1.000000  
50%           24.500000                    3.000000    3.000000    1.000000  
75%           34.250000                    3.000000    4.000000    2.000000  
max           96.000000                    4.000000    5.000000    3.000000